### PR TITLE
Add template for App Service Environment with Az Firewall

### PR DESCRIPTION
**Pull request by jftl6y**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

**Comment by jftl6y** [comment]
Interesting - fixed a link typo in the README.md which kicked off another ci build which is now failing because the ASE it deployed is still there apparently.  How often does the CI environment clean up?

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+                "description": "Name for the NSG attached to the ASE subnet"
+            }
+        },
+        "aseRouteTableName": {
+            "type": "string",
+            "metadata": {
+                "description": "Name of the Route Table attached to the ASE subnet"
+            }
+        },
+        "azureFirewallName": {
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: add a defaultValue

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+            "type": "string",
+            "metadata": {
+                "description": "Name of the Azure Firewall"
+            }
+        },
+        "deployAzureFirewall": {
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: add a defaultValue

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+            "type": "bool",
+            "metadata": {
+                "description": "Toggle whether to deploy the Azure Firewall"
+            }
+        },
+        "azureFirewallRouteTableName": {
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: add a defaultValue

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+            "type": "string",
+            "metadata": {
+                "description": "Name of the Azure Firewall Route Table"
+            }
+        },
+        "azureFirewallSubnetAddressRange": {
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: add a defaultValue

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+            "type": "string",
+            "metadata": {
+                "description": "Address range that will be used by the Azure Firewall Subnet"
+            }
+        },
+        "azureFirewallPublicIP": {
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: add a defaultValue

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+            "type": "string",
+            "metadata": {
+                "description": "Name for the Azure Firewall public IP resource"
+            }
+        },
+        "tags": {
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: add a defaultValue as an empty object

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+        },
+        "deployToAzGov": {
+            "type": "bool",
+            "metadata": {
+                "description": "Toggle whether this will deploy to commercial or Azure Government"
+            }
+        }
+    },
+    "variables": {
+        "vnetID": "[resourceId(resourceGroup().name, 'Microsoft.Network/virtualNetworks', parameters('vnetResourceName'))]",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: remove resourceGroup().name

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+        "nestedTemplateUri": "[if(parameters('deployToAzGov'),concat(parameters('_artifactsLocation'), '/nestedtemplates/vnet.gov.json', parameters('_artifactsLocationSasToken')),concat(parameters('_artifactsLocation'), '/nestedtemplates/vnet.json', parameters('_artifactsLocationSasToken')))]"
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: use the uri function instead of concat (see BP guilde)

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+    },
+    "resources": [
+        {
+            "apiVersion": "2017-05-10",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: update apiVersion

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+            "type": "Microsoft.Resources/deployments",
+            "name": "BuildOrUpdateASENetworking",
+            "resourceGroup": "[resourceGroup().name]",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: remove this property

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+                    },
+                    "deployAzureFirewall": {
+                        "value": "[parameters('deployAzureFirewall')]"
+                    },
+                    "tags": {
+                        "value": "[parameters('tags')]"
+                    }
+                }
+            },
+            "dependsOn": []
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: remove this property

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+        },
+        {
+            "apiVersion": "2015-08-01",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: update apiVersion

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+                "location": "[parameters('location')]",
+                "ipSslAddressCount": 0,
+                "internalLoadBalancingMode": "[parameters('internalLoadBalancingMode')]",
+                "dnsSuffix": "[parameters('dnsSuffix')]",
+                "virtualNetwork": {
+                    "Id": "[variables('vnetID')]",
+                    "Subnet": "[parameters('aseSubnetName')]"
+                }
+            },
+            "dependsOn": [
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: move this above properties

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+                "Microsoft.Resources/deployments/BuildOrUpdateASENetworking"
+            ]
+        },
+        {
+            "type": "Microsoft.web/serverfarms",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: Let's clean up the sort order of these properties, see: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#sort-order-of-properties - won't comment on the rest of them

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+            "tags": {
+                "displayName": "Deploy App Service"
+            },
+            "sku": {
+                "name": "I1",
+                "tier": "Isolated"
+            },
+            "kind": "app",
+            "name": "[parameters('serverfarms_ASE_asp_name')]",
+            "apiVersion": "2016-03-01",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: update apiVersion

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+            ]
+        },
+        {
+            "type": "Microsoft.Web/sites",
+            "tags": {
+                "displayName": "Deploy Web App"
+            },
+            "kind": "app",
+            "name": "[parameters('application_name')]",
+            "apiVersion": "2016-08-01",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: update apiVersion

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+            "location": "West US 2",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: this must be a parameter

Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: has this been tested in fairfax?  (this will fail as written)

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+            "scale": null,
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: remove all null properties

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 19]
```diff
@@ -0,0 +1,326 @@
...
+            "dependsOn": [
+                "[resourceId('Microsoft.Web/serverfarms', parameters('serverfarms_ASE_asp_name'))]",
+                "[resourceId('Microsoft.Web/hostingEnvironments', parameters('aseName'))]"
+            ]
+        },
+        {
+            "type": "microsoft.insights/components",
+            "kind": "web",
+            "name": "[parameters('application_name')]",
+            "apiVersion": "2015-05-01",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: update apiVersion

File: App-Service-Environment-AzFirewall/azuredeploy.parameters.json [hunk 20]
```diff
@@ -0,0 +1,66 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.parameters.json]: remove all parameters that have defaultValue (will be most of them after the template updates)

File: App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json [hunk 21]
```diff
@@ -0,0 +1,666 @@
...
+        },
+        "tags": {
+            "type": "object",
+            "metadata": {
+                "Description": "Resource Tags."
+            }
+        }
+    },
+    "variables": {
+        "nsgId": "[resourceId(resourceGroup().name, 'Microsoft.Network/networkSecurityGroups', parameters('aseNSGName'))]",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: remove resourceGroup().name

File: App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json [hunk 22]
```diff
@@ -0,0 +1,666 @@
...
+                        "properties": {
+                            "description": "Used to manage ASE from public VIP",
+                            "protocol": "*",
+                            "sourcePortRange": "*",
+                            "destinationPortRange": "454-455",
+                            "destinationAddressPrefix": "*",
+                            "access": "Allow",
+                            "priority": 100,
+                            "direction": "Inbound",
+                            "sourcePortRanges": [],
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: remove all empty properties/arrays

File: App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json [hunk 23]
```diff
@@ -0,0 +1,666 @@
...
+                            "destinationPortRanges": [],
+                            "sourceAddressPrefixes": [],
+                            "destinationAddressPrefixes": []
+                        }
+                    }
+                ]
+            }
+        },
+        {
+            "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: update apiVersion

File: App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json [hunk 24]
```diff
@@ -0,0 +1,666 @@
...
+                            "addressPrefix": "0.0.0.0/0",
+                            "nextHopType": "VirtualAppliance",
+                            "nextHopIpAddress": "[variables('azureFireWallPrivateIp')]"
+                        }
+                    }
+                ]
+            }
+        },
+        {
+            "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: update apiVersion

File: App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json [hunk 25]
```diff
@@ -0,0 +1,666 @@
...
+                        "properties": {
+                            "addressPrefix": "0.0.0.0/0",
+                            "nextHopType": "Internet"
+                        }
+                    }
+                ]
+            }
+        },
+        {
+            "apiVersion": "2017-06-01",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: update apiVersion

File: App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json [hunk 26]
```diff
@@ -0,0 +1,666 @@
...
+                        "name": "[parameters('aseSubnetName')]",
+                        "properties": {
+                            "addressPrefix": "[parameters('aseSubnetAddressRange')]",
+                            "networkSecurityGroup": {
+                                "id": "[variables('nsgId')]"
+                            },
+                            "routeTable": {
+                                "id": "[resourceId('Microsoft.Network/routeTables', parameters('aseRouteTableName'))]"
+                            },
+                            "serviceEndpoints": [
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: these should be a parameter since there's nothing in this workload that is using them

File: App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json [hunk 27]
```diff
@@ -0,0 +1,666 @@
+{
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: we need to combine these two templates into a single template - use parameters to set values for the route tables, use a copy loop for the route table definition.

Comment (jftl6y) [reply App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: Here's the problem with this (I have tried to solve this via parameters already and cannot find a more elegant solution):  The default route in the route table is the private IP of the Azure Firewall.  The Azure Firewall IP is derived from the subnet range of the AzureFirewallSubnet, which is itself another parameter.  I'm not sure how to make this dynamic based upon another parameter.  Any suggestions appreciated:

{
                        "name": "DefaultRoute",
                        "properties": {
                            "addressPrefix": "0.0.0.0/0",
                            "nextHopType": "VirtualAppliance",
                            "nextHopIpAddress": "[variables('azureFireWallPrivateIp')]"
                        }
                    }

Comment (bmoore-msft) [reply App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: Not sure I follow - you already have that as a variable (derived from a parameter)

Comment (jftl6y) [reply App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: In the VNet.json, all of the route table entries are for Azure commercial IPs.  In the VNet.gov.json, the routes are for Azure Government IPs.  Normally what I'd do is inject all of those routes as a parameters object EXCEPT I cannot add that DefaultRoute because it relies on the value derived in the variable.

So my options as I see them would be to 1) deploy a super set of the routes for Commercial and Azure Gov (messy if the customer is not in Azure Gov or vice versa) 2) have the user know and understand how to derive the defaultRoute next hop IP address (which would probably get missed much of the time) or 3) have a gov specific template (more maintenance).

I recognize that this is not the most elegant solution, but I'm not sure how else I'd handle it, save drop support for Azure government, which was the impetus for doing most of this.  Again, any suggestions welcome.

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 28]
```diff
@@ -0,0 +1,326 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "location": {
+            "type": "string",
```
[the remark this answers is not in the document]
Comment (jftl6y) [reply App-Service-Environment-AzFirewall/azuredeploy.json]: As this template targets both Commercial and Azure Gov, should I add a value which would make this invalid in one of those environments (e.g. "East US" would break a deploy to Azure Gov)?

[the remark this answers is not in the document]
Comment (bmoore-msft) [reply App-Service-Environment-AzFirewall/azuredeploy.json]: no, the defaultValue for this param must be [resourceGroup().location] - so will work in all clouds

File: App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json (hunk 27 above)
Comment (bmoore-msft) [reply App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: Are you talking about this: https://github.com/Azure/azure-quickstart-templates/pull/5980/files/9889e0547b114888663157fb89b997cfed23a247#diff-dd9ac5fd67628013ee9b9aa3c197b40cR466 - from what I can tell, you already have that authored correctly using the variables which use the parameters you pass in.  If that still doesn't make sense then let's just start working on it and should be easier to debug if something is actually failing.

Comment (bmoore-msft) [reply App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: You'll do something like this:

- pass in an array of address prefixes for the specific cloud
- the first element in the array will be for the default route (0.0.0.0/0)
- create a copy loop to create all the routes
- use the if statement for when copyIndex==0 to add the extra parameters for the defaultRoute, e.g.
  "nextHopType": "[if(equals(copyIndex(), 0), "VirtualAppliance", "Internet")]",
  "nextHopIpAddress": "[if(equals(copyIndex(), 0), variables('azureFireWallPrivateIp'), json('null')]"

I think that's it...

Comment (jftl6y) [reply App-Service-Environment-AzFirewall/nestedtemplates/vnet.gov.json]: I think I get it and will try changing it - thanks.

File: App-Service-Environment-AzFirewall/azuredeploy.json (hunk 19 above)
Comment (jftl6y) [reply App-Service-Environment-AzFirewall/azuredeploy.json]: This is current for microsoft.insights/components 

https://docs.microsoft.com/en-us/azure/templates/microsoft.insights/allversions

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 29]
```diff
@@ -2,13 +2,6 @@
     "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
     "contentVersion": "1.0.0.0",
     "parameters": {
-        "location": {
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: revert this change

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 30]
```diff
@@ -140,27 +143,28 @@
...
+            "defaultValue": [],
             "metadata": {
-                "description": "List of ASE management IPs"
+                "description": "List of ASE management IP addresses"
             }
         }
     },
     "variables": {
-        "vnetID": "[resourceId(resourceGroup().name, 'Microsoft.Network/virtualNetworks', parameters('vnetResourceName'))]",
+        "vnetID": "[resourceId('Microsoft.Network/virtualNetworks', parameters('vnetResourceName'))]",
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: does the vnet have to be in the same resourceGroup?

File: App-Service-Environment-AzFirewall/azuredeploy.json [hunk 31]
```diff
@@ -169,7 +173,7 @@
                 },
                 "parameters": {
                     "location": {
-                        "value": "[parameters('location')]"
+                        "value": "[resourceGroup().location]"
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/azuredeploy.json]: everywhere you have resourceGroup().location, you need to use the parameter

File: App-Service-Environment-AzFirewall/nestedtemplates/vnet.json [hunk 32]
```diff
@@ -613,14 +540,12 @@
                             ]
                         }
                     }
-                ],
-                "natRuleCollections": []
+                ]
             },
             "dependsOn": [
```
Comment (bmoore-msft) [inline App-Service-Environment-AzFirewall/nestedtemplates/vnet.json]: move this above properties and remove resourceGroup().name

File: App-Service-Environment-AzFirewall/azuredeploy.json (hunk 30 above)
Comment (jftl6y) [reply App-Service-Environment-AzFirewall/azuredeploy.json]: No - removed per your earlier comment.
